cmd/sonobuoy/app/operations/cp.go

```diff
@@ -17,10 +17,138 @@ limitations under the License.
 package operations
 
 import (
+	"archive/tar"
 	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/kubectl/cmd"
 )
 
-func CopyResults(path string) error {
-	// pretty straight forwards
-	return errors.New("not implemented")
+// FileSpec defines what and where
+type FileSpec struct {
+	PodNamespace string
+	PodName      string
+	File         string
+}
+
+// CopyResults copies data from src to dst
+func CopyResults(config *rest.Config, clientset internalclientset.Interface, cmderr io.Writer, src, dest FileSpec) error {
+	reader, outStream := io.Pipe()
+	options := &cmd.ExecOptions{
+		StreamOptions: cmd.StreamOptions{
+			In:  nil,
+			Out: outStream,
+			Err: cmderr,
+
+			Namespace: src.PodNamespace,
+			PodName:   src.PodName,
+		},
+
+		// TODO: Improve error messages by first testing if 'tar' is present in the container?
+		Command:  []string{"tar", "cf", "-", src.File},
+		Executor: &cmd.DefaultRemoteExecutor{},
+	}
+
+	go func() {
+		options.Config = config
+		options.PodClient = clientset.Core()
+
+		defer outStream.Close()
+
+		if err := options.Validate(); err != nil {
+			fmt.Println("error validating")
+		}
+		if err := options.Run(); err != nil {
+			fmt.Printf("error running: %v", err)
+		}
+
+	}()
+	prefix := strings.TrimLeft(src.File, "/")
+	prefix = path.Clean(prefix)
+	return untarAll(reader, dest.File, prefix)
+}
+
+func untarAll(reader io.Reader, destFile, prefix string) error {
+	entrySeq := -1
+
+	// TODO: use compression here?
+	tarReader := tar.NewReader(reader)
+	for {
+		header, err := tarReader.Next()
+		if err != nil {
+			if err != io.EOF {
+				return err
+			}
+			break
+		}
+		entrySeq++
+		mode := header.FileInfo().Mode()
+		outFileName := path.Join(destFile, header.Name[len(prefix):])
+		baseName := path.Dir(outFileName)
+		if err := os.MkdirAll(baseName, 0755); err != nil {
+			return err
+		}
+		if header.FileInfo().IsDir() {
+			if err := os.MkdirAll(outFileName, 0755); err != nil {
+				return err
+			}
+			continue
+		}
+
+		// handle coping remote file into local directory
+		if entrySeq == 0 && !header.FileInfo().IsDir() {
+			exists, err := dirExists(outFileName)
+			if err != nil {
+				return err
+			}
+			if exists {
+				outFileName = filepath.Join(outFileName, path.Base(header.Name))
+			}
+		}
+
+		if mode&os.ModeSymlink != 0 {
+			err := os.Symlink(header.Linkname, outFileName)
+			if err != nil {
+				return err
+			}
+		} else {
+			outFile, err := os.Create(outFileName)
+			if err != nil {
+				return err
+			}
+			defer outFile.Close()
+			if _, err := io.Copy(outFile, tarReader); err != nil {
+				return err
+			}
+			if err := outFile.Close(); err != nil {
+				return err
+			}
+		}
+	}
+
+	if entrySeq == -1 {
+		//if no file was copied
+		errInfo := fmt.Sprintf("error: %s no such file or directory", prefix)
```

Comment:
stderr?

Reply:
`Sprintf` is returning a string, so this should be ok

Reply:
oh, sure is. In this case you could use `fmt.Errorf` and combine this line with the next one.